### DISCUSSION TITLE
Custom grep provider

### DISCUSSION THREAD

**Discussion by artemave** [ibhagwan/fzf-lua#1640] [Q&A] [answered]
I have a shell command that outputs vimgrep formatted lines. Looks like this:

```sh
✦ ❯ { git diff master... & git diff } | diff2vimgrep | sort -u
app/controllers/cms/image_uploads_controller.rb:12:1:      response.headers['Cache-Control'] = 'public, max-age=31536000'
app/controllers/cms/image_uploads_controller.rb:13:1:
test/controllers/cms/image_uploads_controller_test.rb:55:1:      assert_redirected_to(rails_representation_url(image_upload.uploaded_image))
test/controllers/cms/image_uploads_controller_test.rb:56:1:
test/controllers/cms/image_uploads_controller_test.rb:57:1:      assert_equal 'max-age=31536000, public', response.headers['cache-control']
```

I'd like to make it a "source" for fzf-lua, but I can't figure out how.

It looks like `grep` provider works with the same kind of vimgrep output:

![image](https://github.com/user-attachments/assets/07d92b2d-42c7-4d94-9643-d80be9c7f231)

But I can't figure out how to substitute the command. I tried `cmd` and `raw_cmd`:

```lua
      vim.keymap.set("n", "<leader>d", function()
        local cmd = "{ git diff master... & git diff } | diff2vimgrep | sort -u"
        fzf_lua.grep {
          search = '',
          cmd = cmd
        }
      end, { desc = "Git diff master" })
```

but I am getting this error:

![image](https://github.com/user-attachments/assets/7070a19f-5855-427d-a83d-c61649aba699)

Using `fzf_exec` works, but gives the raw output, so there is no navigate action, nor highlighting, nor preview. It looks like I need to explain fzf-lua what this output is (again, I should be able to reuse the grep provider options, but I can't figure out how):

![image](https://github.com/user-attachments/assets/6abf712f-3c1d-4040-a89a-c14a8232644f)

Thanks!

**Comment by ibhagwan** [comment]
Grep modifies the command in order to add the search regex at the end of the command, add `debug=true` and you’ll be able to see the generated command  both in the first line of fzf, post the output here and we’ll try to figure it out.

**Reply by artemave** [reply]
> post the output here and we’ll try to figure it out.


With this config:

```lua
      vim.keymap.set("n", "<leader>d", function()
        local cmd = "{ git diff master... & git diff } | diff2vimgrep | sort -u"
        fzf_lua.grep {
          search = '',
          cmd = cmd,
          debug = true,
        }
      end, { desc = "Git diff master" })
```

Here's `messages`:

```
|| [Fzf-lua]: FZF_DEFAULT_COMMAND: VIMRUNTIME='/tmp/.mount_nvimlUMgsh/usr/share/nvim/runtime' '/tmp/.mount_nvimlUMgsh/usr/bin/nvim' -n --headless --clean --cmd 'lua vim.g.did_load_filetypes=1; loadfile([[/home/artem/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/libuv.lua]])().spawn_stdio([==[e2NvbG9yX2ljb25zID0gdHJ1ZSwgY21kID0gInsgZ2l0IGRpZmYgbWFzdGVyLi4uICYgZ2l0IGRpZmYgfSB8IGRpZmYydmltZ3JlcCB8IHNvcnQgLXUgJycgIiwgZyA9IHtfZnpmX2x1YV9zZXJ2ZXIgPSAiL3J1bi91c2VyLzEwMDAvZnpmLWx1YS4xNzM1MTQ0ODk1LjMwNDAyODQuMSIsIF9kZXZpY29uc19wYXRoID0gIi9ob21lL2FydGVtLy5sb2NhbC9zaGFyZS9udmltL2xhenkvbnZpbS13ZWItZGV2aWNvbnMvIn0sIGRlYnVnID0gdHJ1ZSwgcmdfZ2xvYiA9IGZhbHNlLCBnaXRfaWNvbnMgPSBmYWxzZSwgZmlsZV9pY29ucyA9IHRydWV9]==],[==[cmV0dXJuIHJlcXVpcmUoIm1ha2VfZW50cnkiKS5maWxl]==],[==[cmV0dXJuIHJlcXVpcmUoIm1ha2VfZW50cnkiKS5wcmVwcm9jZXNz]==],[==[bmls]==])' 2>&1 || true
|| [Fzf-lua]: fzf cmd: fzf --border 'none' --bind 'ctrl-a:select-all,zero:execute-silent(mkdir '\''/tmp/nvim.artem/nrUN3t/4'\'' && VIMRUNTIME='\''/tmp/.mount_nvimlUMgsh/usr/share/nvim/runtime'\'' '\''/tmp/.mount_nvimlUMgsh/usr/bin/nvim'\'' -n --headless --clean --cmd '\''lua vim.g.did_load_filetypes=1; loadfile([[/home/artem/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/shell_helper.lua]])().rpc_nvim_exec_lua({fzf_lua_server=[[/run/user/1000/fzf-lua.1735144895.3040284.1]], fnc_id=2 , debug=true})'\'' -- ),esc:print(esc)+accept,ctrl-g:print(ctrl-g)+accept,ctrl-q:print(ctrl-q)+accept,alt-Q:print(alt-Q)+accept,alt-q:print(alt-q)+accept,ctrl-s:print(ctrl-s)+accept,enter:print(enter)+accept,ctrl-v:print(ctrl-v)+accept,ctrl-c:print(ctrl-c)+accept,ctrl-t:print(ctrl-t)+accept' --highlight-line --info 'inline-right' --ansi --layout 'reverse' --header ':: <^[[38;2;255;235;205mctrl-g^[[0m> to ^[[38;2;255;64;64mRegex Search^[[0m' --preview 'VIMRUNTIME='\''/tmp/.mount_nvimlUMgsh/usr/share/nvim/runtime'\'' '\''/tmp/.mount_nvimlUMgsh/usr/bin/nvim'\'' -n --headless --clean --cmd '\''lua vim.g.did_load_filetypes=1; loadfile([[/home/artem/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/shell_helper.lua]])().rpc_nvim_exec_lua({fzf_lua_server=[[/run/user/1000/fzf-lua.1735144895.3040284.1]], fnc_id=1 , debug=true})'\'' -- {} {q}' --preview-window 'nohidden:right:0' --prompt 'Rg> ' --print-query --multi --height '100%' > '/tmp/nvim.artem/nrUN3t/6'
```

**Reply by ibhagwan** [reply]
> Here's messages:

That’s not what we need, this is what I meant:
 
![IMG_2839](https://github.com/user-attachments/assets/898cd92a-9e70-43b4-a744-375dc080d631)

**Comment by artemave** [comment]
For the reference (so you could test it yourself), here's [diff2vimgrep](https://github.com/artemave/dotfiles/blob/6635cd0896e87a336fbf64642ba04dfc20614fda/bin/diff2vimgrep#L4) script.

**Reply by ibhagwan** [reply]
Have you seen https://github.com/ibhagwan/fzf-lua/discussions/1598#discussioncomment-11628723?

**Reply by ibhagwan** [reply]
Found the issue that's due to the neovim headless wrapper using `sh -c` to run the command and the `{}` are causing an issue, unsure yet what's the issue sh has with the curly braces but remving these works:

```lua
:lua require("fzf-lua").grep({raw_cmd=[[git diff master... & git diff | diff2vimgrep | sort -u]],debug=true,_base64=
false})
```
![image](https://github.com/user-attachments/assets/8445aa7b-fa65-4023-9155-d95a4e08aca3)

**Reply by ibhagwan** [reply]
Try running this and you'll get a similar error:
```sh
sh -c "{ git diff main... & git diff } | diff2vimgrep | sort -u"
```

![image](https://github.com/user-attachments/assets/0d533067-a058-4f43-bed1-2d7cc2d19af7)

**Reply by ibhagwan** [reply, marked as answer]
Alright solved it, command wasn't POSIX compliant, add a `;` before the 2nd curly brace:
```lua
:lua require("fzf-lua").grep({raw_cmd=[[{ git diff main... & git diff; } | diff2vimgrep | sort -u]],debug=true})
```

![image](https://github.com/user-attachments/assets/47e37932-3c3e-4968-92a6-57dac3db128d)

**Reply by artemave** [reply]
Nice one - it works, thanks!

**Reply by artemave** [reply]
> Have you seen [#1598 (comment)](https://github.com/ibhagwan/fzf-lua/discussions/1598#discussioncomment-11628723)?

This is interesting, I'll give it a spin, thanks!
